### ISSUE TITLE
Min & Max

### ISSUE THREAD

**Issue by sergdort** [RxSwiftCommunity/RxSwiftExt#34]
### Name and description

Min, Max

Propagates `min` or `max` element of the observable sequence based on comparison closure
### Motivation for inclusion

Sometimes it useful to determine `min` or `max` element of the sequence
### Proposed implementation

``` swift
extension ObservableType {
   func edgeElement(isOrderedBefore: (E, E) -> Bool) -> Observable<E> {
      return reduce(Optional<E>.None) { (acum, element) in
         if let current = acum where isOrderedBefore(element, current) {
            return Optional.Some(element)
         }
         if let current = acum where isOrderedBefore(current, element) {
            return acum
         }
         return Optional.Some(element)
         }
         .flatMap { element -> Observable<E> in
            if let some = element {
               return Observable.just(some)
            }
            return Observable.empty()
      }
   }
}

extension ObservableType where E: Comparable {
   func max() -> Observable<E> {
      return edgeElement(>)
   }

   func min() -> Observable<E> {
      return edgeElement(<)
   }
}
```

**Comment by sergdort** [comment]
Not sure about naming though :)

**Comment by icanzilb** [comment]
minLatest() ? :)

**Comment by fpillet** [comment]
What about :

```swift
enum SummaryType {
  case Minimum
  case Maximum
  case Average
  //... etc .. 
}

extension ObservableType {
   public func summaryValue(type: SummaryType) -> Observable<E>
}
```

**Comment by icanzilb** [comment]
this is an interesting idea but I don't think the name of the operator reflects what it'd do - I can't see the emitted value being a "summary". How about "function"?

`myObservable.function(.Minimum)`

**Comment by sergdort** [comment]
Hi, guys.

I also wondering how we would compute `Average ` because it only possible on Numeric types.

So probably will need to introduce a protocol, kind of "DoupleConvertableType"
```swift
protocol  DoupleConvertableType {
    func asDouble() -> Double
}
extension Double: DoupleConvertableType {
    func asDouble() -> Double {
         return self
    }
}
```

**Comment by sergdort** [comment]
In my mind there is also case for `SummaryType ` - `.Count`

**Comment by fpillet** [comment]
I think the best name for this operator would be `compute`. Makes total sense:

```swift
observable.compute(.minimum)
observable.compute(.maximum)
observable.compute(.average)
```

Note that te idea would be that every time a new value comes in, the computation is updated and the latest computed value emitted.

**Comment by freak4pc** [comment]
@fpillet This wouldn't be the behavior of Rx core - which I'm fine with but I want to make sure you're aware of it. I think the behavior you specific here (e.g. on every emission) is more useful but I'm not sure if trying to smoosh them all into one operator is worthwhile. 

I understand _why_ but I'm worried about the discoverability concern here, usually an operator does one thing, and here you have an operator that does several different things. Not sure how I feel about that 🤔

**Comment by fpillet** [comment]
@freak4pc what about `updatedSummary(.minimum)` ?

**Comment by freak4pc** [comment]
Even more confusing IMO 

I'm not arguing with the compute name - I'm arguing with putting everything in a single operator... 
I'm not sure it's good or bad TBH - I think people are used to looking for the operator name from the docs, which might be confusing. 

On the other hand if it uses the same name but a different implementation (e.g. not only on completion, but accumulative), that might also be confusing. 

tl;dr I think `.compute(.minimum)` or `.math(.minium)` would be find if we got for a single operator. If we go for separate we can just the standard names perhaps?

**Comment by fpillet** [comment]
I see what you mean. Since it's not the same computation with different options but different computation, it makes sense to use different operator names.

I was trying to synthesize a set of behavior but maybe it's a better choice to be explicit then.

**Comment by freak4pc** [comment]
So, we can name them average, min, and max - the only risk here is that if Krunoslav ever agrees to add the default Rx implementations - we would have a conflict. We can deal with this later to be honest. 

Also, we might want to support both behaviors 

`Observable.min(cumulative: Bool = true)`

Setting to false will have the "Rx-standard" behavior (e.g. compute on completion) 

What do you think ?

**Comment by freak4pc** [comment]
I'm closing this for now due to inactivity - If you have any more comments on this subject, feel free to comment or ping me directly and I'll re-open it :) Thanks !
